ksql-engine/src/main/java/io/confluent/ksql/function/udaf/complex/AverageUdaf.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.function.udaf.complex;
+
+import io.confluent.ksql.function.udaf.TableUdaf;
+import io.confluent.ksql.function.udaf.UdafDescription;
+import io.confluent.ksql.function.udaf.UdafFactory;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+
+@UdafDescription(name = "average",
+    description = "Returns the average value of the column computed as the sum divided by the"
+        + "count. Applicable only to numeric types.")
+public final class AverageUdaf {
+
+  private AverageUdaf() {
+  }
+
+  private static final Schema SCHEMA = SchemaBuilder.struct().optional()
+      .field("SUM", Schema.OPTIONAL_INT64_SCHEMA)
+      .field("COUNT", Schema.OPTIONAL_INT64_SCHEMA)
+      .build();
+
+  private static TableUdaf<Long, Struct, Double> average() {
+    return new TableUdaf<Long, Struct, Double>() {
+
+      @Override
+      public Struct initialize() {
+        return new Struct(SCHEMA).put("SUM", 0L).put("COUNT", 0L);
+      }
+
+      @Override
+      public Struct aggregate(final Long newValue,
+                              final Struct aggregate) {
+
+        return new Struct(SCHEMA)
+            .put("SUM", (Long)aggregate.get("SUM") + newValue)
+            .put("COUNT", (Long)aggregate.get("COUNT") + 1);
+
+      }
+
+      @Override
+      public Struct merge(final Struct agg1,
+                          final Struct agg2) {
+
+        return new Struct(SCHEMA)
+            .put("SUM", (Long)agg1.get("SUM") + (Long)agg2.get("SUM"))
+            .put("COUNT", (Long)agg1.get("COUNT") + (Long)agg2.get("COUNT"));
+      }
+
+      @Override
+      public Struct undo(final Long valueToUndo,
+                         final Struct aggregate) {
+
+        return new Struct(SCHEMA)
+            .put("SUM", (Long)aggregate.get("SUM") - valueToUndo)
+            .put("COUNT", (Long)aggregate.get("COUNT") - 1);
+      }
+
+      @Override
+      public Double map(final Struct aggregate) {
+        return aggregate.getInt64("SUM") / ((double)aggregate.getInt64("COUNT"));
```

Comment:
classic edge case! need to check for `COUNT == 0` 😉 make sure to add a unit test for that